CALCITE-4408 Adding an operand type checker for the Oracle substring …

jamesstarr opened the pull request:
…opereator(James Starr)

julianhyde commented on core/src/test/java/org/apache/calcite/sql/test/SqlOperatorBaseTest.java:
```diff
@@ -2186,6 +2186,13 @@ protected static Calendar getCalendarNotTooNear(int timeUnit) {
     checkConcat2Func(tester(SqlLibrary.ORACLE));
   }
 
+  @Test void testSubString() {
```
It looks as if this function was never tested. Thanks for fixing the hole.

Rename the test method to `testOracleSubstrFunction` and move it next to `testSubstringFunction`.

Test a few more cases - e.g. indexes out of bounds, and null values, and CHAR argument.

vlsi commented:
> I.e. treat Oracle's function as syntactic sugar

That would not work if downstream Oracle DB has a functional-based index with `SUBSTRING`.
If Calcite transforms `SUBSTRING` into `SUBSTR`, then the query won't use the index.

jamesstarr commented:
From what I have read, oracles substr has a different behavior then the standard sql substr.  Particularly, that is can return null if a 0 or a negative length is passed in, it translates the 0 start index to 1, and if the start index is negative then the start indexes from the end.

@julianhyde, are you suggesting simply aliasing substr to substring?  Because that will not work due to the above reasons.   If you are suggesting using an api that would intercept the arguments at compile time, that would not work due to function arguments, as @vlsi has already stated.  Are you suggesting using a different API to implement the above logic, then could you please point me to class, so I can implement it in the correct place.

julianhyde commented:
Let's just suppose for a minute that Oracle's SUBSTR function had the same semantics as standard SUBSTRING, just different syntax. Then it is not a goal to keep it intact through Calcite's planning process. Calcite's philosophy is to convert to a minimal language. That creates opportunities for simplification/optimization.

If Oracle is the target, then we should translate SUBSTRING back to SUBSTR when we generate Oracle SQL. But it is not necessarily the target.

julianhyde commented:
If Oracle's `SUBSTR` has very different semantics, treat it as a different function. Consider calling it `ORACLE_SUBSTR` or something so that no one gets confused. Adding a method to `SqlFunctions` may make things a lot more concise.

julianhyde commented:
Also, correct the javadoc "It has similar semantics to standard SQL's {@link SqlStdOperatorTable#SUBSTRING} function but different syntax". That is a wrong assumption.

jamesstarr commented:
@julianhyde, I renamed SqlLibraryOperators.SUBSTR to SqlLibraryOperators.ORACLE_SUBSTR and I have updated the comment.

If I understand you other comments, you are saying the implementation would be straight forward to add a function SqlFunctions and then bind it in BuiltInMethod, so it works for the unit tests.  And the benefit of doing this would be that it is more concise and if you translated it back sql, the native function would still used.  The disadvantages are that non-oracle targets would have to implement the function themselves and simplification/optimization could not occur since it is not part of the calcites canonical form.

julianhyde commented:
@jamesstarr I've thought about this some more, and I now think your approach of expanding Oracle SUBSTR calls to standard SUBSTRING calls + CASE is better than adding a new method to SqlFunctions.

jamesstarr commented:
If this looks good, can I get this merged?

vlsi commented on core/src/main/java/org/apache/calcite/sql2rel/StandardConvertletTable.java:
```diff
@@ -1506,6 +1508,83 @@ private SqlNode getCastedSqlNode(SqlNode argInput, RelDataType varType,
+            startLessThanZero, startToSubstr.apply(negativeIndexStart),
+            startToSubstr.apply(start));
+      default:
+        throw new AssertionError();
```
Please add the relevant message to clarify why this branch must not be reachable.

julianhyde replied:
I don't think that's necessary.

jamesstarr replied:
Done

jamesstarr commented:
@julianhyde @vlsi Why was commit that allowed the substr to be correctly executed on different sql engines not included?

julianhyde commented:
@jamesstarr The purpose of this bug (as I see it, and as evidenced by the test cases in this PR) is to make sure that the Oracle `SUBSTR` function returns the correct results when run in Calcite's native mode.

I acknowledge that I ignored most of your most recent commit. I don't think that it improved things. Your change was based based on the idea that Calcite's standard `SUBSTRING` operator would have different semantics depending on the SQL engine that it is implemented on. That is a deeply flawed assumption.

Let's log a follow-up bug (or bugs) to make sure that `SUBSTRING`, and Oracle `SUBSTR`, function correctly when we generate SQL against other engines, and make sure that the fixes include test cases.

I have already logged https://issues.apache.org/jira/browse/CALCITE-4427 to restore standards-compliant behavior to the SUBSTRING operator and I am working on it.
